Introducing ros2 parameters and name remapping

wayneparrott opened the pull request:
This PR introduces ROS2 parameter and name remapping support.
Resolves Issues #414, #415, #416 and #421.

wayneparrott committed:
Initial ROS2 parameter and remapping support.

Parameter Use-cases:
  Declare Parameters
  Override Declared Parameters
  Override Declared Parameters from yaml parameters file
  List Parameters
  Describe Parameters
  Get Parameters
  Set Parameters
  Set Parameters Atomically

Remap Use-cases:
	Remap node name
	Remap node namespace name
	Remap publisher topic name
	Remap service name

1. Name remapping and paramter overrides w/ yaml file support is
delegated to rcl api.

2. index#init() modified to accept an argv array with --ros-args
commandline arguments.

3. index#createNode() accepts NodeOptions for programmatic defined
parameter overrides.

4. Added parameter.js with
	ParameterType
	Parameter
	ParameterDescription
	FloatingPointRange
	IntegerRange

5. Added is-close module dependency for use in the FloatingPointRange
class.

6. New rcl_bindings functions:
  getParameterOverrides()
  getTopic()
  getServiceName()

7. Node.js changes:

	added instance vars:
  	_parameters
  	_parameterDescriptors
  	_parameterService
  	_parameterEventPublisher

	init() function changes:
  	sets up parameter-overrides
  	optionally declares parameters from parameter-overrides
  	starts parameterService
  	creates parameterEventPublisher

	Node methods:
  	getParameterOverrides()
  	hasParameter()
  	declareParameter()
  	declareParameters()
  	undeclareParameter()
  	getParameter()
  	getParameters()
  	hasParameterDescriptor()
  	getParameterDescriptor()
  	getParameterDescriptors()
  	setParameter()
  	setParameters()
  	setParameterAtomically()
  	setParametersAtomically()
  	_getNativeParameterOverrides()
  	_valdateParameters()

8. Tests files:
	test-parameters.js
	test-parameter-service.js
	test-remapping.js

9. Examples:
	parameter-declaration-example.js
	parameter-override-example.js

Commit:
Change the ROS2 packaging used for CIs

Currently, the ROS2 packages (last stable) used on macOS fails consistently
and those packages were built several months ago.

This patch changes to the last successful ones which can reflect the
ROS2 changes in time.

Fix #None

Commit:
Add project name and version number to jsdoc

This patch incorporates the package.json to add project's name and
version number on the homepage when generating jsdoc

Fix #None

Commit:
Bump to v0.12.0

Fix #None

Commit:
Update release information for v0.12.0

Fix #None

wayneparrott committed:
Initial ROS2 parameter and remapping support.

Parameter Use-cases:
  Declare Parameters
  Override Declared Parameters
  Override Declared Parameters from yaml parameters file
  List Parameters
  Describe Parameters
  Get Parameters
  Set Parameters
  Set Parameters Atomically

Remap Use-cases:
	Remap node name
	Remap node namespace name
	Remap publisher topic name
	Remap service name

1. Name remapping and paramter overrides w/ yaml file support is
delegated to rcl api.

2. index#init() modified to accept an argv array with --ros-args
commandline arguments.

3. index#createNode() accepts NodeOptions for programmatic defined
parameter overrides.

4. Added parameter.js with
	ParameterType
	Parameter
	ParameterDescription
	FloatingPointRange
	IntegerRange

5. Added is-close module dependency for use in the FloatingPointRange
class.

6. New rcl_bindings functions:
  getParameterOverrides()
  getTopic()
  getServiceName()

7. Node.js changes:

	added instance vars:
  	_parameters
  	_parameterDescriptors
  	_parameterService
  	_parameterEventPublisher

	init() function changes:
  	sets up parameter-overrides
  	optionally declares parameters from parameter-overrides
  	starts parameterService
  	creates parameterEventPublisher

	Node methods:
  	getParameterOverrides()
  	hasParameter()
  	declareParameter()
  	declareParameters()
  	undeclareParameter()
  	getParameter()
  	getParameters()
  	hasParameterDescriptor()
  	getParameterDescriptor()
  	getParameterDescriptors()
  	setParameter()
  	setParameters()
  	setParameterAtomically()
  	setParametersAtomically()
  	_getNativeParameterOverrides()
  	_valdateParameters()

8. Tests files:
	test-parameters.js
	test-parameter-service.js
	test-remapping.js

9. Examples:
	parameter-declaration-example.js
	parameter-override-example.js

wayneparrott committed:
Parameter integration fixes.

minggangw commented on src/rcl_bindings.cpp:
```diff
@@ -65,11 +68,23 @@ NAN_METHOD(Init) {
+  // preprocess argc & argv
+  v8::Local<v8::Array> jsArgv = v8::Local<v8::Array>::Cast(info[1]);
+  int argc = jsArgv->Length();
+  const char *argv[argc];
```
I found the detailed information https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2131?view=vs-2019, I think it's a ms compiler-specific error.

Also, |argv| should not be const as it will be assigned in L82

wayneparrott replied:
Fixed with 586cd4fa72652fdacf4758f20c88616bed0eabcf

wayneparrott commented:
<disregard> 
I modified the Publish#topic and Service#serviceName getters to return the base name similar to the previous versions but with remapping support now.

minggangw commented:
Please rebase on the `develop` branch and 602a5dfdcf0ef483f302644de4839a797e7dcacd has resolved the failure on CircleCI.

wayneparrott committed:
Merge develop onto parameter PR

Conflicts:
	src/rcl_bindings.cpp

mattrichard commented on lib/node.js:
```diff
@@ -15,30 +15,138 @@
     this._services = [];
     this._timers = [];
     this._guards = [];
     this._name = name;
```
Should we remove `_name` now to avoid accidental usage? I think it's being used in a couple places in this class that should now use `this.name()` instead.

wayneparrott replied:
yes, I think name should be removed.

wayneparrott replied:
done

mattrichard commented on lib/node.js:
```diff
@@ -47,6 +155,40 @@ class Node {
     }
     this._namespace = namespace;
```
ditto. Remove this._namespace?

wayneparrott replied:
My vote is yes. @minggangw may have some historical perspective that we need to consider.

minggangw replied:
We can remove the |_namespace| and |_name| now as we have the binding methods which can get them through rcl. So please remove them. Thanks for implementing the series of param related features, I may spend some time to have a look and will give comments when it's finished, thanks!

wayneparrott replied:
done.

mattrichard commented on index.js:
```diff
@@ -100,23 +104,30 @@ let rcl = {
     let node =  new rclnodejs.ShadowNode();
-
-    node.init(nodeName, namespace);
+	  node.handle = handle;
```
`node.handle` is assigned in two places now: here and below the debug statement. Maybe just pass it into init instead and let the Node assign it internally?

wayneparrott replied:
good catch. I'm not sure how I goofed that up. But clearly I did. I'll commit a fix. thx.

wayneparrott replied:
fixed in 9e4b732

wayneparrott replied:
fixed in 9e4b732

mattrichard commented on index.js:
```diff
@@ -126,9 +137,10 @@ let rcl = {
   /**
   * Init the module.
   * @param {Context} context - The context, default is Context.defaultContext().
+  * @param {String[]} argv - Process commandline arguments.
```
nit: should be lowercase 's' on string. Idk if jsdoc distinguishes between 'string' and 'String'

wayneparrott replied:
fixed in 9e4b732

wayneparrott replied:
fixed in 9e4b732

wayneparrott committed:
Fixes errant duplicate node.handle assigment & jsdoc string type.

wayneparrott committed:
Fixes errant duplicate node.handle assigment & jsdoc string type.

wayneparrott committed:
ignore ros pkg dirs: install/, build/ & log/

wayneparrott committed:
Merge new ref libs.

minggangw commented on lib/node.js:
```diff
@@ -15,30 +15,138 @@
+ * node instantiation process.  
+ * @class
+ */
+class NodeOptions {
```
Shall we move `NodeOpions` to a separate file? We have had a lot of functions here.

wayneparrott replied:
agree, I refactored NodeOptions to its own class, see commit 52acc9dc26ccd993751ac8cab8033e21a8c3a0ab

minggangw commented on lib/node.js:
```diff
@@ -15,30 +15,138 @@
+   * @param {boolean} aBool - True indicates for a node to start it's parameter_service.
+   *    Default value is true.
+   */
+  set startParameterServices(aBool) {
```
I suggest we rename `aBool` in order to be meaningful.

wayneparrott replied:
fixed

minggangw commented on lib/node.js:
```diff
@@ -15,30 +15,138 @@
+   * parameter-overrides.
+   * @param {boolean} aBool - 
+   */
+  set automaticallyDeclareParametersFromOverrides(aBool) {
```
ditto

wayneparrott replied:
fixed

minggangw commented on lib/node.js:
```diff
@@ -47,6 +155,40 @@ class Node {
+        }
+        this._parameterOverrides.set(parameter.name, parameter);
+      }
+    }
```
Do we need to handle the `else` when `options.parameterOverrides` is not Array as an exception?

wayneparrott replied:
When I refactored NodeOptions to its own file I also added checking to ensure that its _parameterOverrides property is always assigned an array of Parameters; the default value is empty array []. See commit 52acc9d

minggangw commented on lib/node.js:
```diff
@@ -538,6 +680,498 @@ class Node {
+   * @return {Parameter[]} - The declared parameters.
+   */
+  declareParameters(parameters, descriptors=[], ignoreOverrides = false) {
+    
```
This bank line can be removed.

wayneparrott replied:
fixed. Also went thru my recent code additions attempting to normalize use of vertical whitespace to match your guidance for reduced vert ws..

minggangw commented on lib/node.js:
```diff
@@ -538,6 +680,498 @@ class Node {
+   * @return {undeined} -
+   */
+  undeclareParameter(name) {
+    if (! this.hasParameter(name)) return;
```
additional space after `!`

wayneparrott replied:
fixed by jsbeautify

minggangw commented on lib/node.js:
```diff
@@ -538,6 +680,498 @@ class Node {
+   *    descriptors to find or null indicating to return all declared descriptors.
+   * @return {ParameterDescriptor[]} - The parameter descriptors. 
+   */
+  getParameterDescriptors(names=[]) {
```
space around `=`

wayneparrott replied:
fixed by jsbeautify.

minggangw commented on lib/node.js:
```diff
@@ -538,6 +680,498 @@ class Node {
+   * @param {Parameter[]} parameters - The parameters to set.
+   * @return {rclnodejs.rcl_interfaces.msg.SetParameterResult[]} - A list of SetParameterResult, one for each parameter that was set.
+   */
+  setParameters(parameters=[]) {
```
ditto

wayneparrott replied:
fixed by jsbeautify.

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+   * @param {Range} [range] - An optional IntegerRange or FloatingPointRange. 
+   */
+  constructor(name, type = ParameterType.PARAMETER_NOT_SET, 
+    description='no description', readOnly=false, range=null) {
```
ditto, add space around `=`

wayneparrott replied:
fixed by jsbeautify.

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+      throw TypeError('Incompatible Range');
+    }
+
+    this._range = ranget;
```
typo?

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+      throw TypeException('Expected instance of Range.');
+    }
+
+    if (! this.range.isValidType(this.type)) {
```
additional space after `!` L371-380

wayneparrott replied:
tightened up vertical spacing by removing blank lines.

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+   * @return {boolean} - True if value satisfies the range; false otherwise.
+   */
+  inRange(value) {
+    
```
Please delete this line

wayneparrott replied:
fixed

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+  if (!value) return ParameterType.PARAMETER_NOT_SET;
+  if (typeof value === 'boolean') return ParameterType.PARAMETER_BOOL;
+  if (typeof value == 'string') return ParameterType.PARAMETER_STRING;
+  if (typeof value == 'number') return ParameterType.PARAMETER_DOUBLE;
```
shall we use `===`?

wayneparrott replied:
updated all '==' to '==='

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+ * @return {boolean} - True if value is a valid ParameterType; false otherwise.
+ */
+function validType(parameterType) {
+    
```
ditto

wayneparrott replied:
fixed

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+ * @return {boolean} - True if value can be represented by type.
+ */
+function validValue(value, type) {
+
```
ditto

wayneparrott replied:
done

minggangw commented on lib/parameter.js:
```diff
@@ -0,0 +1,774 @@
+        values.reduce((compatible, val) => 
+          compatible && validValue(val, arrayElementType), 
+        true);
+    }
```
Would you please put it into a private function?

wayneparrott replied:
done, refactored to _isArrayParameterType(type)

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+  PARAMETER_SEPARATOR } = require('./parameter.js');
+
+
+class ParameterService {
```
Missing jsdoc comments?

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+  // type ListParametersResult = {
+  //      names: string[],
+  //      prefixes: string[]
+  //    };
```
is it still useful?

wayneparrott replied:
removed.

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+    const result = msg.result;
+
+    for (const paramName of parameterNames) {
+
```
delete this line

wayneparrott replied:
done

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+      if (paramName.includes(PARAMETER_SEPARATOR)) {
+        prefixedNames.push(paramName);
+        continue;
+
```
delete this line

wayneparrott replied:
done

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+      } else if (request.prefixes.length > 0) {
+        
+        for (const prefix of request.prefixes) {
+          
```
delete this line

wayneparrott replied:
done

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+            let fullPrefix = paramName.split(PARAMETER_SEPARATOR);
+            fullPrefix.pop();
+            fullPrefix = fullPrefix.join(PARAMETER_SEPARATOR);
+            if (! result.prefixes.includes(fullPrefix)) {
```
ditto

wayneparrott replied:
fixed

minggangw commented on lib/parameter_service.js:
```diff
@@ -0,0 +1,255 @@
+            if (! result.prefixes.includes(fullPrefix)) {
+              result.prefixes.push(fullPrefix);
+            }
+            if (! request.prefixes.includes(prefix)) {
```
ditto

wayneparrott replied:
fixed

minggangw commented on package-lock.json:
```diff
@@ -0,0 +1,2925 @@
+          "requires": {
+            "emoji-regex": "^7.0.1",
+            "is-fullwidth-code-point": "^2.0.0",
+            "strip-ansi": "^5.1.0"
```
I think we could exclude `package-lock.json` to be submitted

wayneparrott replied:
agree

minggangw replied:
I think we could omit this file here :)

minggangw reviewed (commented):
Wow, such a big PR. I did a quick review and find some problems. I'd like to give some comments quickly and continue to finish the detailed part. I think we should use [js-beautify](https://www.npmjs.com/package/js-beautify) to format the JS code, I believe it can save our time a lot. I am going to add the js-beautify.

wayneparrott committed:
Merged jsbeautify, deep-equal, obsolete action removal

wayneparrott committed:
Refactored NodeOptions to its own file & updated dependencies.

wayneparrott committed:
Merge remote-tracking branch 'upstream/develop' into develop

wayneparrott committed:
Refactored NodeOptions type from node.*.ts, updated tests

wayneparrott committed:
format, jsdoc & parameter descriptor range tests improvements.

wayneparrott committed:
fixes format & lint errors, remove node._name & node._namespace cleanup

- remove node._name & node._namespace
- renamed aBool parameters with contextual meaningful name
- fixes lint format errors

wayneparrott committed:
Merge LTS Erbium changes.

koonpeng commented on src/rcl_bindings.cpp:
```diff
@@ -65,12 +68,24 @@ NAN_METHOD(Init) {
   THROW_ERROR_IF_NOT_EQUAL(RCL_RET_OK,
-                           rcl_init(0, nullptr, &init_options, context),
+                           rcl_init(argc, argc > 0 ? argv: nullptr,
+                                    &init_options, context),
```
several potential problems i see here:
* i dont see `argv` being freed any where.
* `strdup` is not a standard C function, the microsoft docs says that using this will produce a compiler warning in msvc https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup?view=vs-2019. It's probably better to either have a `#if` macro that use different functions depending on platform, or use malloc/new + strcpy/std::copy/memcpy.
* The `rcl_init` docs says that "rcl specific arguments are removed", I'm not sure if it means actually modifying argv and `delete`ing the rcl args, if so we need to be careful when cleaning up `argv` so we don't end up with a double free.
* `Nan::Utf8String()` may fail and contain a nullptr, we should probably handle it before we go into a segfault.

minggangw replied:
>strdup is not a standard C function, the microsoft docs says that using this will produce a compiler warning in msvc https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup?view=vs-2019. It's probably better to either have a #if macro that use different functions depending on platform, or use malloc/new + strcpy/std::copy/memcpy.

I prefer std::strcpy

>The rcl_init docs says that "rcl specific arguments are removed", I'm not sure if it means actually modifying argv and deleteing the rcl args, if so we need to be careful when cleaning up argv so we don't end up with a double free.

Yes, we have to take care of the memory management here. I took a look at the implementation and don't find `argv` is freed.

wayneparrott replied:
> The rcl_init docs says that "rcl specific arguments are removed", I'm not sure if it means actually modifying argv and deleteing the rcl args, if so we need to be careful when cleaning up argv so we don't end up with a double free.

The rcl_init doc contradicts itself by stating in 1 place that the ros args are remove from argv as you identify above. But in the main paragraph of the function it states that ros arguments are parsed and not removed from argv. The latter behavior of not modifying argv makes much better sense to me - but that's beside the point here. Looking deeper into the rcl#init() code confirms that argv is being copied into the local Context impl. Given that there is no concern for a redundant free() calls.

wayneparrott replied:
* replaced strdup() with snprintf(). When I used strcpy() I got a compiler warning recommending to use snprintf(). 
* added logic when argc == 0. Should be rare occurance.
* argv is now freed.

wayneparrott replied:
I added parameter checking in rclnodejs.init() that detects invalid argv values.

mattrichard commented on lib/node.js:
```diff
@@ -533,11 +580,492 @@ class Node {
+   * 
+   * Readonly parameters can not be undeclared or updated. 
+   * @param {string} name - Name of parameter to undeclare.
+   * @return {undeined} -
```
Typo in 'undefined'

wayneparrott replied:
fixed

mattrichard commented on lib/node.js:
```diff
@@ -533,11 +580,492 @@ class Node {
+    }
+
+    // create ParameterEvent
+    const parameterEvent = new(loader.loadInterface(PARAMETER_EVENT_MSG_TYPE));
```
Unless I'm missing something, since loadInterface returns a class type shouldn't this be `new loader.loadInterface(PARAMETER_EVENT_MSG_TYPE)()`. Do you not need the () in ES6?

wayneparrott replied:
Good question. Originally I had the snippet as you shared which results in this runtime error: `TypeError: loader.loadInterface is not a constructor`.  After seeing this error msg I added the outer parens to force it to work.

Marking resolved. Please reopen for additional discussion as I thought I knew the operator precedence but missed on this one initially.

mattrichard commented on lib/node.js:
```diff
@@ -533,11 +580,492 @@ class Node {
+
+    if (newParameters.length > 0) {
+      // eslint-disable-next-line camelcase
+      parameterEvent.new_parameters =
```
You can avoid needing to disable eslint rules by using the [] for property assignments. E.g. `parameterEvent['new_parameter'] = ...`. If you prefer to stick with what you currently have that's fine with me too.

wayneparrott replied:
I can go either way. Sticking with disabling eslint rule atm to get this merge done. Will discuss w/ you and @minggangw later and will gladly revise eslint rule disabling is discouraged for such cases.

mattrichard reviewed (commented):
I took one last look and found a few very minor things. Otherwise, it LGTM.

wayneparrott committed:
Fixed whitespace issue.

wayneparrott committed:
Formatted with Prettier

wayneparrott committed:
Merge Pretty formatted codebase into parameters

Conflicts:
	index.js
	lib/node.js
	test/test-destruction.js
	test/test-extra-destroy-methods.js
	test/test-node.js
	test/utils.js
	types/index.d.ts
	types/node.d.ts

wayneparrott committed:
Fix jsdoc typo.

wayneparrott committed:
Fixes eslint camelCase issue by using bracket property access

wayneparrott committed:
Fix test cases to account for node param event publisher & service.

koonpeng commented on src/rcl_bindings.cpp:
```diff
@@ -83,6 +98,153 @@ NAN_METHOD(Init) {
+      reinterpret_cast<rcl_context_t*>(context_handle->ptr());
+
+  rcl_arguments_t *parsed_args = &(context->global_arguments);
+  rcl_params_t * params = NULL;
```
`params` was never cleaned up

wayneparrott replied:
fixed.

koonpeng reviewed (commented):
i just did a quick look, so it seems parameter service is not in rcl. Hopefully it will be added in the future and we can use the rcl implementation so we don't need to chase any API changes in future ros releases.

minggangw commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# rclnodejs - ROS2 Client Library for JavaScript [![npm](https://img.shields.io/npm/v/rclnodejs.svg)](https://www.npmjs.com/package/rclnodejs)[![Coverage Status](https://coveralls.io/repos/github/RobotWebTools/rclnodejs/badge.svg?branch=develop)](https://coveralls.io/github/RobotWebTools/rclnodejs?branch=develop)[![npm](https://img.shields.io/npm/dm/rclnodejs)](https://www.npmjs.com/package/rclno ...
+# rclnodejs - ROS2 Client Library for JavaScript [![npm](https://img.shields.io/npm/v/rclnodejs.svg)](https://www.npmjs.com/package/rclnodejs)[![Coverage Status](https://coveralls.io/repos/github/RobotWebTools/rclnodejs/badge.svg?branch=develop)](https://coveralls.io/github/RobotWebTools/rclnodejs?branch=develop)[![npm](https://img.shields.io/npm/dm/rclnodejs)](https://www.npmjs.com/package/rclno ...
```
Please rebase with `develop`, it seems that you delete the prettier badge accidentally.

wayneparrott replied:
fixed.

minggangw commented on index.js:
```diff
@@ -73,6 +75,9 @@ let rcl = {
   _initialized: false,
   _nodes: [],
 
+  /** {@link Context} class */
```
Thanks for adding this comment block.

minggangw commented on jsbeautifyrc:
```diff
@@ -0,0 +1,20 @@
+    "unescape_strings": false,
+    "wrap_line_length": 120,
+    "operator_position": "before-newline"
+}
```
Please rebase the latest `develop` as this configure file has been deleted.

wayneparrott replied:
fixed

minggangw commented on lib/node.js:
```diff
@@ -607,13 +653,520 @@ class Node {
+
+    if (newParameters.length > 0) {
+      parameterEvent['new_parameters'] = newParameters.map((
+        parameter // eslint-disable-line camelcase
```
Do we still need // eslint-disable-line camelcase, the same with L1047 & 1052

wayneparrott replied:
not needed now. removed.

minggangw commented on lib/node_options.js:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * Set startParameterServices.
+   * @param {boolean} startFlag - True indicates for a node to start it's parameter_service.
+   *    Default value is true.
```
I'm afraid that `Default value is true.` may cause misunderstanding (it's not the `startFlag` actually).

wayneparrott replied:
fixed

minggangw commented on src/rcl_bindings.cpp:
```diff
@@ -83,6 +98,153 @@ NAN_METHOD(Init) {
+
+parameters = array<Node>;
+*/
+static v8::Local<v8::Object> wrapParameters(rcl_params_t *parsed_args, v8::Isolate* isolate) {  // NOLINT(whitespace/line_length)
```
I suggest we use `Nan` classes as much as possible, as the changes of v8 could bring incompatibility.

- [Nan::New()](https://github.com/nodejs/nan/blob/master/doc/new.md#api_nan_new) to create javascript objects.
- [Nan::Set()](https://github.com/nodejs/nan/blob/master/doc/maybe_types.md#api_nan_set) to set the elements of an array.

wayneparrott replied:
done. Thanks for recommendation.

minggangw reviewed (commented):
Please remove the binary files `test/cpp/add_two_ints_client` and `test/cpp/listener`, I think we could land this huge PR since the comments are addressed, thanks for your great patience to implement the param features.

>i just did a quick look, so it seems parameter service is not in rcl. Hopefully it will be added in the future and we can use the rcl implementation so we don't need to chase any API changes in future ros releases.

Absolutely@koonpeng, I have aligned with @wayneparrott we implement a JS version of parameter service quickly and the 2nd step is to replace it with rcl bindings while keeping the js interface the same as it is.

wayneparrott committed:
Parameter code clean up

- src/rcl_bindings_cpp
 use Nan::Set/Get/New
 free argv
 free param struct

- removed files
 test/cpp/add_two_ints_client
 test/cpp/listener

- fixed silly merge goofs
 node.js
 index.js
 README
 jsbeautifyrc

- lib/node_options.js
 improved jsdoc

wayneparrott committed:
Removed package-lock.json

wayneparrott committed:
Merge branch 'develop' into develop

wayneparrott committed:
Fix init() with invalid argv values

wayneparrott committed:
Merged reformatted files

wayneparrott commented:
A future TODO which I'll open as an issue for the future is to be able to quickly access the non-ros2 commandline parameters. There is a an rcl api for this. I just decided it is lower priority atm and would like to see the core parameter and name remapping use-cases wrapped up.

wayneparrott committed:
Added libary rcl_yaml_param_parser

minggangw commented on package.json:
```diff
@@ -18,7 +18,7 @@
     "lint": "eslint --max-warnings=0 index.js types/*.d.ts scripts lib example rosidl_gen rosidl_parser test benchmark/rclnodejs && node ./scripts/cpplint.js",
     "postinstall": "node scripts/generate_messages.js",
-    "format": "prettier --trailing-comma es5 --single-quote --write \"{.,{lib,rosidl_gen,rostsd_gen,rosidl_parser,types,example,test}/**}/*.{js,md,ts}\""
+    "format": "prettier --trailing-comma es5 --single-quote --write '{.,lib,rosidl_gen,rostsd_gen,rosidl_parser,types,example,test}/*.{js,md,ts}'"
```
It seems that you overwrite the configuration of prettier :)

minggangw commented:
Hi @koonpeng do you have any other suggestions/comments? If not, I am going to merge this PR when the CIs pass.

wayneparrott committed:
Fix reintro prettier omitted from previous parameters merge.

koonpeng commented:
shouldn't we commit package-lock.json? It would help CI to get reproducible builds, but if we are not using that then it should probably be added to .gitignore.

Other than that everything LGTM.

minggangw reviewed (approved):
The builds are not cached on CIs currently and I take a loot at https://docs.npmjs.com/cli/ci.html, it seems that we have to execute `npm ci` to make use of `package-lock.json`. I think we could add it to  .gitignore.

Alright, let's land it now! Thanks for implementing and reviewing the PR!
